static/graph-layout-core.ts

```diff
@@ -0,0 +1,742 @@
+        // just naively iterates.
+        enum EventType {
+            Edge = 0,
+            Block = 1,
+        }
+        type Event = {
+            blockIndex: number;
+            edgeIndex: number;
+            row: number;
+            type: EventType;
+        };
+        const events: Event[] = [];
+        for (const [i, block] of this.blocks.entries()) {
+            events.push({
+                blockIndex: i,
+                edgeIndex: -1,
+                row: block.row,
+                type: EventType.Block,
+            });
+            for (const [j, edge] of block.edges.entries()) {
+                events.push({
+                    blockIndex: i,
+                    edgeIndex: j,
+                    row: Math.max(block.row + 1, this.blocks[edge.dest].row),
+                    type: EventType.Edge,
+                });
+            }
+        }
+        // Sort by row (max(src row, target row) for edges), edge row n is before block row n
+        events.sort((a: Event, b: Event) => {
+            if (a.row === b.row) {
+                return a.type - b.type;
+            } else {
+                return a.row - b.row;
+            }
+        });
+        //
+        const blockedColumns = Array(this.columnCount + 1).fill(-1);
+        for (const event of events) {
+            if (event.type === EventType.Block) {
+                const block = this.blocks[event.blockIndex];
+                blockedColumns[block.col + 1] = block.row;
+            } else {
+                const source = this.blocks[event.blockIndex];
+                const edge = source.edges[event.edgeIndex];
+                const target = this.blocks[edge.dest];
+                const sourceColumn = source.col + 1;
+                const targetColumn = target.col + 1;
+                const topRow = Math.min(source.row + 1, target.row);
+                if (blockedColumns[sourceColumn] < topRow) {
+                    // use column under source block
+                    edge.mainColumn = sourceColumn;
+                } else if (blockedColumns[targetColumn] < topRow) {
+                    // use column of the target
+                    edge.mainColumn = targetColumn;
+                } else {
+                    const leftCandidate =
+                        sourceColumn -
+                        1 -
+                        blockedColumns
+                            .slice(0, sourceColumn)
+                            .reverse()
+                            .findIndex(v => v < topRow);
+                    const rightCandidate = sourceColumn + blockedColumns.slice(sourceColumn).findIndex(v => v < topRow);
+                    // hamming distance
+                    const distanceLeft =
+                        Math.abs(sourceColumn - leftCandidate) + Math.abs(sourceColumn - rightCandidate);
+                    const distanceRight =
+                        Math.abs(targetColumn - leftCandidate) + Math.abs(targetColumn - rightCandidate);
```

Comment:
This is wrong. Should be:
```
distanceLeft = |source-left | + |target - left|
```
not
```
distanceLeft = |source-left | + |target - right|
```
This is probably the cause for loop edges going the long way  around and up in this  image:
![image](https://user-images.githubusercontent.com/7101031/180614612-c89bf8e3-19be-400a-8c4d-233b89863527.png)

Reply:
Ope thank you!

Reply:
The loop edge is going the wrong way here too
![image](https://user-images.githubusercontent.com/51220084/180614986-84a7c984-9c63-4f8e-a694-23c776bf818c.png)

The left and right distance should be the same, not sure how to introduce the necessary information to be able to chose the path with fewer conflicts here

Reply:
Try adding the code which talks about  figure 8 shape. It's supposed to deal with slightly different case but might help in this one as well.